src/common/numbers.cpp

```diff
@@ -0,0 +1,311 @@
+/*
+ * Copyright (c) 2020 Emanuel Machado da Silva
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "numbers.h"
+
+#include "common/strings.h"
+#include "compat/compare.h"
+
+#include <algorithm>
+#include <cassert>
+#include <cinttypes>
+#include <cmath>
+#include <utility>
+
+using namespace Puzzles::Numbers;
+
+Number::Number(std::string value) : numerator(std::move(value)), denominator("1"), positive(true) {
+  // TODO: assert there are no invalid characters
+  if (numerator[0] == '-') {
+    numerator = numerator.substr(1);
+    positive = false;
+  }
+  numerator = trimLeadingView(numerator, '0');
+  if (numerator.empty()) {
+    numerator = "0";
+  }
+}
+
+Number::Number(intmax_t value) : Number(std::to_string(value)) {}
+
+Number::Number(intmax_t _n, uintmax_t _d) : Number(_n >= 0 ? _n : _n * -1, _d, _n >= 0) {}
+
+Number::Number(uintmax_t numerator, uintmax_t denominator, bool positive)
+    : numerator(std::to_string(numerator)), denominator(std::to_string(denominator)), positive(positive) {}
+
+Number::Number(std::string numerator, std::string denominator, bool positive)
+    : numerator(std::move(numerator)), denominator(std::move(denominator)), positive(positive) {}
+
+template <typename iterator>
+inline char valueAndAdvance(iterator *it) {
+  char value = **it;
+  std::advance(*it, 1);
+  return value;
+}
+
+compat::strong_ordering compareIntegers(const std::string &left, const std::string &right) {
+  if (left.length() != right.length()) {
```

Comment:
this will work with leading zeros? 🤔

Reply:
It won't, but that's not an issue. I actually designed the class in a way that there will never be any leading zeros here. Great catch, though! I'm adding an assertion here to make that more explicit! Thanks 😄

Reply:
Actually, on second thought... I changed the class to store zero as an empty string (previously I had a special case where zero was the only number with a leading zero). It feels less "special-casey" that way, like I don't have to treat the "special case of zero" so much. Unfortunately, I had a very optimized `operator==(intmax_t)` that didn't agree with that, so that got special cased instead... _sigh_ .
Oh well, still happier with the new result, please check the new commit d9876f6.

Reply:
Oops, there was a small error left uncaught, so I amended it. Please look at commit c280e8a instead 😬